DDDEastAnglia/Controllers/AccountController.cs

```diff
@@ -141,70 +161,78 @@ public ActionResult ChangePassword(LocalPasswordModel model)
+
+            ViewBag.ReturnUrl = returnUrl;
+            ViewBag.DisableRemoveButtons = hasLocalAccount || usersExternalLogins.Count() > 1;
+            var viewModels = allExternalLogins.Select(loginMethodViewModelBuilder.Build);
+            return PartialView("_ExternalLoginsPartial", viewModels);
+        }
+        
+        [AllowAnonymous]
+        [ChildActionOnly]
+        public ActionResult ExternalLoginMethods()
+        {
+            var externalLogins = externalLoginsProvider.GetAllAvailable();
+            var viewModels = externalLogins.Select(loginMethodViewModelBuilder.Build);
+            return PartialView("_ExternalLoginMethodsPartial", viewModels);
         }
 
         [AllowAnonymous]
         [ChildActionOnly]
-        public ActionResult ExternalLoginsList(string returnUrl)
+        public ActionResult ExternalLoginButtons(string returnUrl)
         {
             ViewBag.ReturnUrl = returnUrl;
-            return PartialView("_ExternalLoginsListPartial", OAuthWebSecurity.RegisteredClientData);
+            var externalLogins = externalLoginsProvider.GetAllAvailable();
+            var viewModels = externalLogins.Select(loginMethodViewModelBuilder.Build);
+            return PartialView("_ExternalLoginButtonsPartial", viewModels);
         }
 
         [HttpPost]
         [ValidateAntiForgeryToken]
-        public ActionResult DisassociateLogin(string provider, string providerUserId)
+        public ActionResult DisassociateLogin(string name, string provider, string providerUserId)
         {
-            string ownerAccount = OAuthWebSecurity.GetUserName(provider, providerUserId);
+            string ownerAccount = oAuthClientInfo.GetUserName(provider, providerUserId);
             string message = null;
 
-            // Only disassociate the account if the currently logged in user is the owner
             if (ownerAccount == User.Identity.Name)
             {
-                // Use a transaction to prevent the user from deleting their last login credential
-                using (var scope = new TransactionScope(TransactionScopeOption.Required, new TransactionOptions { IsolationLevel = IsolationLevel.Serializable }))
-                {
-                    bool hasLocalAccount = OAuthWebSecurity.HasLocalAccount(WebSecurity.GetUserId(User.Identity.Name));
+                bool hasLocalAccount = oAuthClientInfo.HasLocalAccount(WebSecurity.GetUserId(User.Identity.Name));
+                var externalLogins = oAuthClientInfo.GetAccountsFromUserName(User.Identity.Name);
 
-                    if (hasLocalAccount || OAuthWebSecurity.GetAccountsFromUserName(User.Identity.Name).Count > 1)
-                    {
-                        OAuthWebSecurity.DeleteAccount(provider, providerUserId);
-                        scope.Complete();
-                        message = "The external login has been removed successfully.";
-                    }
+                if (hasLocalAccount || externalLogins.Count > 1)
```

Comment:
This logic appears to be duplicated - or at least there's something *very* similar higher up. Can we extract a method to put a name on it?